From 2038549f9b0a0b5c4031c9cc8413fc697a8032b1 Mon Sep 17 00:00:00 2001
From: "Mazlan, Hazwan Arif" <hazwan.arif.mazlan@intel.com>
Date: Tue, 18 Nov 2025 23:23:22 +0800
Subject: [PATCH 2/8] kmsro: Add iris renderonly support

Enable using iris for KMS renderonly.

v2: let kmsro_drm_screen_create() be invoked by kmsro driver.(Tina)
v3: fix icd build error (Dongwon & Alyssa)
v4: Rebased toward v25.3.4 (Hazwan)

Upstream PR: https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/9592/
Upstream-status: Innapropriate

OSV: https://jira.devtools.intel.com/browse/PKT-19618

Signed-off-by: Tina Zhang <tina.zhang@intel.com>
Signed-off-by: Dongwon Kim <dongwon.kim@intel.com>
Signed-off-by: Mazlan, Hazwan Arif <hazwan.arif.mazlan@intel.com>
---
 meson.build                                         | 1 +
 src/gallium/auxiliary/pipe-loader/meson.build       | 3 +++
 src/gallium/auxiliary/pipe-loader/pipe_loader_drm.c | 5 +++++
 src/gallium/auxiliary/target-helpers/drm_helper.h   | 7 ++++++-
 src/gallium/winsys/iris/drm/iris_drm_public.h       | 4 ++--
 src/gallium/winsys/iris/drm/iris_drm_winsys.c       | 6 +++---
 src/gallium/winsys/kmsro/drm/kmsro_drm_winsys.c     | 6 ++++++
 src/loader/loader.c                                 | 4 ++--
 8 files changed, 28 insertions(+), 8 deletions(-)

diff --git a/meson.build b/meson.build
index cc901ec629e..83146d83a50 100644
--- a/meson.build
+++ b/meson.build
@@ -247,6 +247,7 @@ with_gallium_kmsro = system_has_kms_drm and [
   with_gallium_v3d,
   with_gallium_vc4,
   with_gallium_zink,
+  with_gallium_iris,
 ].contains(true)
 
 _vulkan_drivers = get_option('vulkan-drivers')
diff --git a/src/gallium/auxiliary/pipe-loader/meson.build b/src/gallium/auxiliary/pipe-loader/meson.build
index 2bda1d28872..78023b36914 100644
--- a/src/gallium/auxiliary/pipe-loader/meson.build
+++ b/src/gallium/auxiliary/pipe-loader/meson.build
@@ -45,6 +45,9 @@ endif
 if with_gallium_ethosu
   renderonly_drivers_c_args += '-DGALLIUM_ETHOSU'
 endif
+if with_gallium_iris
+  renderonly_drivers_c_args += '-DGALLIUM_IRIS'
+endif
 
 libpipe_loader_static = static_library(
   'pipe_loader_static',
diff --git a/src/gallium/auxiliary/pipe-loader/pipe_loader_drm.c b/src/gallium/auxiliary/pipe-loader/pipe_loader_drm.c
index ee6500e9e3c..0c4b4499b25 100644
--- a/src/gallium/auxiliary/pipe-loader/pipe_loader_drm.c
+++ b/src/gallium/auxiliary/pipe-loader/pipe_loader_drm.c
@@ -388,6 +388,9 @@ pipe_loader_get_compatible_render_capable_device_fds(int kms_only_fd, unsigned i
 #endif
 #if defined GALLIUM_VC4
       "vc4",
+#endif
+#if defined GALLIUM_IRIS
+     "i915",
 #endif
    };
 
@@ -396,10 +399,12 @@ pipe_loader_get_compatible_render_capable_device_fds(int kms_only_fd, unsigned i
    is_platform_device = (dev->type == PIPE_LOADER_DEVICE_PLATFORM);
    pipe_loader_release(&dev, 1);
 
+#ifndef GALLIUM_IRIS
    /* For display-only devices that are not on the platform bus, we can't assume
     * that any of the rendering devices are compatible. */
    if (!is_platform_device)
       return NULL;
+#endif
 
    /* For platform display-only devices, we try to find a render-capable device
     * on the platform bus and that should be compatible with the display-only
diff --git a/src/gallium/auxiliary/target-helpers/drm_helper.h b/src/gallium/auxiliary/target-helpers/drm_helper.h
index 236b68f8c0e..4e336132002 100644
--- a/src/gallium/auxiliary/target-helpers/drm_helper.h
+++ b/src/gallium/auxiliary/target-helpers/drm_helper.h
@@ -53,6 +53,7 @@ const struct drm_driver_descriptor descriptor_name = {         \
 #undef GALLIUM_ASAHI
 #undef GALLIUM_ROCKET
 #undef GALLIUM_ETHOSU
+#undef GALLIUM_IRIS
 #endif
 
 #ifdef GALLIUM_I915
@@ -507,9 +508,13 @@ const driOptionDescription kmsro_driconf[] = {
 #ifdef GALLIUM_LIMA
       #include "lima/driinfo_lima.h"
 #endif
+#ifdef GALLIUM_IRIS
+      #include "iris/driinfo_iris.h"
+#endif
 };
 DRM_DRIVER_DESCRIPTOR(kmsro, kmsro_driconf, ARRAY_SIZE(kmsro_driconf))
-
+#elif defined(GALLIUM_IRIS)
+DRM_DRIVER_DESCRIPTOR(kmsro, iris_driconf, ARRAY_SIZE(iris_driconf))
 #else
 DRM_DRIVER_DESCRIPTOR_STUB(kmsro)
 #endif
diff --git a/src/gallium/winsys/iris/drm/iris_drm_public.h b/src/gallium/winsys/iris/drm/iris_drm_public.h
index e8e49758024..afb640cc854 100644
--- a/src/gallium/winsys/iris/drm/iris_drm_public.h
+++ b/src/gallium/winsys/iris/drm/iris_drm_public.h
@@ -32,7 +32,7 @@ struct pipe_screen *
 iris_drm_screen_create(int drm_fd, const struct pipe_screen_config *config);
 
 struct pipe_screen *
-iris_drm_screen_create_renderonly(struct renderonly *ro,
-                              const struct pipe_screen_config *config);
+iris_drm_screen_create_renderonly(int fd, struct renderonly *ro,
+                                  const struct pipe_screen_config *config);
 
 #endif /* IRIS_DRM_PUBLIC_H */
diff --git a/src/gallium/winsys/iris/drm/iris_drm_winsys.c b/src/gallium/winsys/iris/drm/iris_drm_winsys.c
index 011311ecd50..8cfcc5f7a6e 100644
--- a/src/gallium/winsys/iris/drm/iris_drm_winsys.c
+++ b/src/gallium/winsys/iris/drm/iris_drm_winsys.c
@@ -40,12 +40,12 @@ iris_drm_screen_create(int fd, const struct pipe_screen_config *config)
 }
 
 struct pipe_screen *
-iris_drm_screen_create_renderonly(struct renderonly *ro,
-                              const struct pipe_screen_config *config)
+iris_drm_screen_create_renderonly(int fd, struct renderonly *ro,
+                                  const struct pipe_screen_config *config)
 {
    struct iris_screen *pscreen;
 
-   pscreen = (struct iris_screen *)iris_screen_create(os_dupfd_cloexec(ro->gpu_fd), config);
+   pscreen = (struct iris_screen *)iris_screen_create(os_dupfd_cloexec(fd), config);
    if (!pscreen)
       return NULL;
 
diff --git a/src/gallium/winsys/kmsro/drm/kmsro_drm_winsys.c b/src/gallium/winsys/kmsro/drm/kmsro_drm_winsys.c
index deb847d994e..e676c633572 100644
--- a/src/gallium/winsys/kmsro/drm/kmsro_drm_winsys.c
+++ b/src/gallium/winsys/kmsro/drm/kmsro_drm_winsys.c
@@ -33,6 +33,7 @@
 #include "panfrost/drm/panfrost_drm_public.h"
 #include "lima/drm/lima_drm_public.h"
 #include "asahi/drm/asahi_drm_public.h"
+#include "iris/drm/iris_drm_public.h"
 #include "xf86drm.h"
 
 #include "pipe/p_screen.h"
@@ -125,6 +126,11 @@ struct pipe_screen *kmsro_drm_screen_create(int kms_fd,
          */
          ro->create_for_resource = renderonly_create_gpu_import_for_resource;
          screen = vc4_drm_screen_create_renderonly(ro->gpu_fd, ro, config);
+#endif
+      } else if (strcmp(render_dev_name, "i915") == 0) {
+#if defined(GALLIUM_IRIS)
+	 ro->create_for_resource = renderonly_create_kms_dumb_buffer_for_resource;
+	 screen = iris_drm_screen_create_renderonly(ro->gpu_fd, ro, config);
 #endif
       }
 
diff --git a/src/loader/loader.c b/src/loader/loader.c
index 0567beb3dee..46e26c81b63 100644
--- a/src/loader/loader.c
+++ b/src/loader/loader.c
@@ -249,8 +249,8 @@ loader_open_render_node_platform_devices(const char * const drivers[],
    for (i = 0; i < num_devices; i++) {
       device = devices[i];
 
-      if ((device->available_nodes & (1 << DRM_NODE_RENDER)) &&
-          (device->bustype == DRM_BUS_PLATFORM)) {
+      if ((device->available_nodes & (1 << DRM_NODE_RENDER)) /*&&
+          (device->bustype == DRM_BUS_PLATFORM)*/ ) {
          drmVersionPtr version;
 
          fd = loader_open_device(device->nodes[DRM_NODE_RENDER]);
-- 
2.43.0

